activerecord/lib/active_record/encryption/cipher.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module Encryption
+    # The algorithm used for encrypting and decrypting +Message+ objects.
+    #
+    # It uses AES-256-GCM. It will generate a random IV for non deterministic encryption (default)
+    # or derive an initialization vector from the encrypted content for deterministic encryption.
```

Comment:
Have you considered blind indexing instead of a deterministic IV? CipherSweet does this ([initial blog post](https://paragonie.com/blog/2019/01/ciphersweet-searchable-encryption-doesn-t-have-be-bitter), [open sourced PHP implementation](https://ciphersweet.paragonie.com/)) to prevent data leakage.

I started on a [Ruby port](https://github.com/bdewater/ciphersweet-ruby) but with the pandemic I didn't have the energy to complete a Rails integration. Later I found out another gem has implemented a similar (but not exactly the same IIRC) scheme: https://github.com/ankane/blind_index

Reply:
Hi @bdewater, I did, but preferred to favor this version because it doesn't make you add new columns.  The initialization vector is calculated as a SHA-256 hash of the text to encrypt and the secret, which should be offer a similar protection against leakage. Said that, it might be an interesting option to add in the future.

Reply:
@bdewater I am curious, can you think of advantages for the blind index approach?

Reply:
The three CipherSweet links probably do a better job explaining than I could (especially [this one](https://paragonie.com/blog/2017/05/building-searchable-encrypted-databases-with-php-and-sql)), but in short blind indexing prevents an attacker from learning that user 1 and user 2 have the same encrypted data because they have the same deterministic ciphertext. Given a large enough data set of deterministic ciphertexts, I could even do frequency analysis on e.g. names and guess what the plaintexts are.

Blind indexing prevents this: for example user 1, 2, 8 and 24 share the same blind index value even though only user 1 and 2 share the same plaintext (8 and 24 are both different) because it truncates the data. That does mean the database returns these 4 rows, which all need to be decrypted and the non-matching rows discarded, as a trade-off. The [CipherSweet PHP docs have a section](https://ciphersweet.paragonie.com/php/blind-index-planning) on how to determine the right blind index configuration preserve performance and not leak data.

I'm not a cryptographer by any means, but a (hopefully competent) implementer that does his best to avoid the common footguns... so seeing IVs and ciphertext being deterministic looks quite concerning. The aforementioned link also calls out "Insecure modes include ECB mode and CBC mode with a static (or empty) IV" (and using GCM doesn't make a difference here).

As an aside: wanted to say I really appreciate the effort you put in! I worked on a similar approach using the Attributes API for encryption but I never got around to what RelationQueries/CoreQueries does in this PR. Nice! 👏

Reply:
> The aforementioned link also calls out "Insecure modes include ECB mode and CBC mode with a static (or empty) IV" (and using GCM doesn't make a difference here).

Notice that the IV we use is not fixed! It derived deterministically from the plain text to encrypt. Ideally we would be using [AES-GCM-SIV](https://tools.ietf.org/html/rfc8452) to achieve what we want: have the same ciphertext for the same plaintext and key, but it wasn't available in the current OpenSSL Ruby bindings, so we are essentially doing the same manually: we derive the IV as a deterministic function of the key and plaintext (HMAC-SHA256). This was actually one of the points the security firm helped us to fix back in the day: we were fixing the IV which was a major flaw. So different plaintexts use different initialization vectors, but we use the same IV for the same ciphertext.

Thanks for the additional explanations and resources 🙏.

Reply:
Ah, I missed that, thanks for pointing it out! That should take care of data confidentiality issue I think, although it's more leaky then CipherSweet's design. [This article](https://blog.cryptographyengineering.com/2019/02/11/attack-of-the-week-searchable-encryption-and-the-ever-expanding-leakage-function/) makes a compelling argument leakage is not avoidable anyway, so we're back to making trade-offs.